### PR TITLE
Handle transactions without crypto transfers in transaction REST API

### LINKED ISSUE #1295: Transactions REST API returns no result for system accounts
**Detailed Description**
System accounts 0.0.2 and 0.0.50 (confirm if others) have recently been changed to have their fees waived. The REST API `/api/v1/transactions/{transactionId}` assumes that there are crypto transfers when constructing its query. Since there are none, it returns no results.

**Actual Behavior**
1. `curl -s "http://mainnet.mirrornode.hedera.com/api/v1/transactions/0.0.2-1605733260-000000000"`
```json
{"_status":{"messages":[{"message":"Not found"}]}}
```
**Expected Behavior**
The transaction is returned for system accounts.

**Environment:**
- Version: 0.23.1

**Additional Context**

### PR DESCRIPTION
**Pull request by xin-hedera**
**Detailed description**:

- `account.id` can be either the payer account id of a transaction or the entity id in the crypto transfer, if the query does not use credit/debit type filter
- refactor mysql style query to postgre style query conversion so duplicate query param value concat is not needed
- add new transaction table index

**Which issue(s) this PR fixes**:
Fixes #1295 

**Special notes for your reviewer**:

**Checklist**
- [ ] Documentation added
- [x] Tests updated

### CODE REVIEW

File: hedera-mirror-rest/transactions.js [hunk 1]
```diff
@@ -223,7 +270,10 @@ const reqToSql = function (req) {
   const resultTypeQuery = utils.parseResultParams(req, 't.result');
   const transactionTypeQuery = utils.getTransactionTypeQuery(parsedQueryParams);
   const {query, params, order, limit} = utils.parseLimitAndOrderParams(req);
-  const sqlParams = accountParams.concat(tsParams).concat(tsParams).concat(params);
+
+  // accountQuery and tsQuery will appear twice in the inner query, one for transaction table and one for
+  // crypto_transfer table. So concat their params twice here to match the query.
```
Comment (xin-hedera) [inline hedera-mirror-rest/transactions.js:275]: the comment no longer applies, will remove in the next commit

File: hedera-mirror-rest/transactions.js [hunk 2]
```diff
@@ -169,6 +171,22 @@ const buildWhereClause = function (...conditions) {
...
+ *
+ * @param {String} query - they mysql query
+ * @param {String} prefix - the prefix for the named parameters
+ * @return {String} - The converted query
+ */
+const convertToNamedQuery = function (query, prefix) {
+  let index = 0;
+  return query.replace(/\?/g, () => {
+    const namedParam = `?${prefix}${index}`;
+    index += 1;
```
Comment (ijungmann) [inline hedera-mirror-rest/transactions.js:185]: Showing my JS ignorance a little, I assume `index` is initialized only on the first call and this continues to increment, instead of constantly resetting to 0, correct?

**Review by ijungmann — CHANGES_REQUESTED** [review]
Still reviewing, here's what I have so far.

File: hedera-mirror-rest/transactions.js [hunk 3]
```diff
@@ -195,23 +213,62 @@ const getTransactionsInnerQuery = function (
...
+
+  const transactionAccountQuery = namedAccountQuery.replace(/ctl\.entity_id/g, 't.payer_account_id');
+  const transactionWhereClause = buildWhereClause(
+    transactionAccountQuery,
+    namedTsQuery,
+    resultTypeQuery,
+    transactionTypeQuery
+  );
+  const ctlJoinClause =
+    (resultTypeQuery || transactionTypeQuery) && 'JOIN transaction AS t ON ctl.consensus_timestamp = t.consensus_ns';
```
Comment (ijungmann) [inline hedera-mirror-rest/transactions.js:249]: Interesting, did not know about this.

**Review by ijungmann — CHANGES_REQUESTED** [review]
Couple more things, mostly smaller questions.  Tests look good.

File: hedera-mirror-rest/transactions.js [hunk 4]
```diff
@@ -169,6 +171,22 @@ const buildWhereClause = function (...conditions) {
...
 };
 
+/**
+ * Convert parameters to the named format.
+ *
+ * @param {String} query - they mysql query
+ * @param {String} prefix - the prefix for the named parameters
+ * @return {String} - The converted query
+ */
+const convertToNamedQuery = function (query, prefix) {
```
[the remark this answers is not in the document]
Comment (xin-hedera) [reply hedera-mirror-rest/transactions.js:181]: AFAIK, the mysql style query was written at the beginning when the choice of db was mysql. that's why for account/balance/transaction, the query needs to be translated to pg style.

I intend to limit the scope of "converting mysql style ? placeholder to named parameter" to where it's indeed needed, for instance, the transactions inner query since the same REST query param is used in multiple places so in the previous PR `tsParams` were concat twice.

Eventually, we should remove the translation layer and directly compose pg style query as in topicmessages and tokens.

File: hedera-mirror-rest/transactions.js [hunk 5]
```diff
@@ -195,23 +213,62 @@ const getTransactionsInnerQuery = function (
...
+    SELECT coalesce(t.consensus_ns,ctl.consensus_timestamp) AS consensus_timestamp
+    FROM (
+        SELECT consensus_ns
+        FROM transaction AS t
+        ${transactionWhereClause}
+        ORDER BY consensus_ns ${order}
+        ${namedLimitQuery}
+    ) AS t
+    FULL OUTER JOIN (
+        SELECT DISTINCT ctl.consensus_timestamp AS consensus_timestamp
```
[the remark this answers is not in the document]
Comment (xin-hedera) [reply hedera-mirror-rest/transactions.js:262]: a crypto transfer always has an associated transaction. previously, the reverse is also true that a transaction always has a crypto transfer list.

similarly, a token transfer must belong to a transaction.

File: hedera-mirror-rest/utils.js [hunk 6]
```diff
@@ -255,14 +254,14 @@ const parseOperatorAndValueFromQueryParam = (paramValue) => {
   if (splitItem.length === 1) {
     // No operator specified. Just use "eq:"
     return {op: opsMap.eq, value: splitItem[0]};
-  } else if (splitItem.length === 2) {
+  }
+  if (splitItem.length === 2) {
```
[the remark this answers is not in the document]
Comment (xin-hedera) [reply hedera-mirror-rest/utils.js:258]: this is auto adjusted by IDEA eslint plugin based on some rule.

File: hedera-mirror-rest/utils.js [hunk 7]
```diff
@@ -355,7 +354,8 @@ const parseCreditDebitParams = (parsedQueryParams, columnName) => {
   return parseParams(parsedQueryParams[constants.filterKeys.CREDIT_TYPE], (op, value) => {
     if (value === 'credit') {
       return [`${columnName} > 0`, []];
-    } else if (value === 'debit') {
+    }
+    if (value === 'debit') {
```
[the remark this answers is not in the document]
Comment (xin-hedera) [reply hedera-mirror-rest/utils.js:358]: this is auto adjusted by IDEA eslint plugin based on some rule.

File: hedera-mirror-rest/transactions.js (hunk 5 above)
[the remark this answers is not in the document]
Comment (ijungmann) [reply hedera-mirror-rest/transactions.js:262]: That makes sense, but then I'm not understanding why a FULL OUTER JOIN is needed, as opposed to a LEFT OUTER JOIN.

File: hedera-mirror-rest/transactions.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (ijungmann) [reply hedera-mirror-rest/transactions.js:181]: That makes sense, thanks for the details.

File: hedera-mirror-rest/balances.js [hunk 8]
```diff
@@ -91,7 +91,7 @@ const getBalances = async (req, res) => {
       ${query}`;
 
   const sqlParams = tsParams.concat(accountParams).concat(pubKeyParams).concat(balanceParams).concat(params);
-  const pgSqlQuery = utils.convertMySqlStyleQueryToPostgres(sqlQuery, sqlParams);
+  const pgSqlQuery = utils.convertMySqlStyleQueryToPostgres(sqlQuery);
```
[the remark this answers is not in the document]
Comment (xin-hedera) [reply hedera-mirror-rest/balances.js:94]: good catch

File: hedera-mirror-rest/transactions.js (hunk 5 above)
[the remark this answers is not in the document]
Comment (xin-hedera) [reply hedera-mirror-rest/transactions.js:262]: oh, it's partially because I made it generic for all REST query params.

To be concise, it's because of the `account.id` filter. with the change, it can be either `transaction.payer_account_id` or `crypto_transfer.entity_id`. It's possible that for `account.id` A, the transactions of which A paid fee for and the transactions with A in the crypto transfer have entries not in the other. So we need a full outer join.

[the remark this answers is not in the document]
Comment (ijungmann) [reply hedera-mirror-rest/transactions.js:262]: Ah, got it.  Interesting that that one param causes so much havoc, that makes sense though./

[the remark this answers is not in the document]
Comment (xin-hedera) [reply hedera-mirror-rest/transactions.js:268]: made some changes to the query to split the generic query into two cases, with and without account query. also added comments based on your feedback

File: hedera-mirror-rest/transactions.js [hunk 9]
```diff
@@ -195,23 +213,72 @@ const getTransactionsInnerQuery = function (
...
-            ORDER BY consensus_timestamp ${order}
-         ) AS ctl
-    ON t.consensus_ns = ctl.consensus_timestamp
-    ${whereClause}
-    ORDER BY t.consensus_ns ${order}
-    ${limitQuery}`;
+  const ctlTsQuery = namedTsQuery.replace(/t\.consensus_ns/g, 'ctl.consensus_timestamp');
+  let ctlWhereClause;
+
+  if (creditDebitQuery) {
```
Comment (Nana-EC) [inline hedera-mirror-rest/transactions.js:225]: I get the logic but this section is a little hard to follow and I worry that its complexity might contribute to challenges in managing it.
Do you think there's a way to simplify or split it up into clearer functions.

**Review by Nana-EC — CHANGES_REQUESTED** [review]
Looking good.
Just a suggestion on simplifying it where possible

File: hedera-mirror-rest/transactions.js (hunk 9 above)
Comment (xin-hedera) [reply hedera-mirror-rest/transactions.js]: made an attempt, let me know if it's more maintainable

File: hedera-mirror-rest/__tests__/specs/accounts-10-account-transactions.spec.json [hunk 10]
```diff
@@ -83,6 +92,20 @@
   "responseStatus": 200,
   "responseJson": {
     "transactions": [
+      {
```
Comment (Nana-EC) [inline hedera-mirror-rest/__tests__/specs/accounts-10-account-transactions.spec.json:95]: Can you add another spec for account transaction with no transfers to make sure that's picked up.

**Review by Nana-EC — COMMENTED** [review]
Looks good, additional suggestion for a spec for the accounts/<accountId> spec for a transaction with no

**Review by ijungmann — APPROVED** [review]
LGTM

**Review by Nana-EC — APPROVED** [review]
LGTM

**Review by ijungmann — APPROVED** [review]

**Review by steven-sheehy — APPROVED** [review]
